### PR TITLE
mon: block osd pool mksnap for fs pools

### PR DESCRIPTION
**Pull request by mchangir**
Fixes: https://tracker.ceph.com/issues/59552
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by mchangir** [comment]
jenkins test make check

**Comment by mchangir** [comment]
jenkins test make check

**Comment by mchangir** [comment]
> @mchangir Needs a _bit_ more explanation in the commit message.

ack; added a detailed note in the commit message

**Comment by vshankar** [comment]
LGTM. Need @rzarzynski review though.

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -13033,6 +13033,12 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
     if (pp->snap_exists(snapname.c_str())) {
       ss << "pool " << poolstr << " snap " << snapname << " already exists";
     } else {
+      if (const auto& fsmap = mon.mdsmon()->get_fsmap(); fsmap.pool_in_use(pool)) {
```
Comment (rzarzynski) [inline src/mon/OSDMonitor.cc:13036]: From the commit description:

> Pool level snaps and mon managed snaps can't co-exist (...)

Isn't this a thinko? Per https://github.com/ceph/ceph/pull/47753#discussion_r953292185 I started thinking it's about _Pool level snaps and **FS level snaps**_.

File: qa/tasks/cephfs/test_snapshots.py [hunk 2]
```diff
@@ -550,6 +550,16 @@ def test_disallow_monitor_managed_snaps_for_fs_pools(self):
         with self.assertRaises(CommandFailedError):
             self.fs.rados(["mksnap", "snap2"], pool=self.fs.get_metadata_pool_name())
 
+        with self.assertRaises(CommandFailedError):
+            test_pool_name = self.fs.get_data_pool_name()
+            base_cmd = f'osd pool mksnap {test_pool_name} snap3'
+            self.run_cluster_cmd(base_cmd)
+
+        with self.assertRaises(CommandFailedError):
```
Comment (rzarzynski) [inline qa/tasks/cephfs/test_snapshots.py:558]: ACK.

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (mchangir) [reply src/mon/OSDMonitor.cc:13036]: yes; this is about FS level snaps and Pool level snaps
its just the nomenclature that @gregsfortytwo has used to refer to it (AFAIUI)

but looks like the phrasing should be changed to _FS level snaps and Pool level snaps can't coexist_

Comment (mchangir) [reply src/mon/OSDMonitor.cc:13036]: phrasing changed to _"Pool-level and fs-level snapshots ..._"

**Comment by mchangir** [comment]
rebased as well ...

**Comment by vshankar** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins test windows

**Comment by vshankar** [comment]
@yuriw please pick this up for main branch test runs.

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomcBVxlgRvT1782-wip-yuri5-testing-2023-06-28-1515-old-wip-yuri5-testing-2023-06-21-0750-old-wip-yuri5-testing-2023-06-16-1012-old-wip-yuri5-test

**Pull request merged**